RavenDB-19718: Reimplementation of SortingMatch

redknightlois opened the pull request:
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-19718
https://issues.hibernatingrhinos.com/issue/RavenDB-19755/StressTests.Corax.OrderByMultiSortingTests.OrderByWithRepetitions
https://issues.hibernatingrhinos.com/issue/RavenDB-19706

### Additional description
Order-by statements would not return proper ordering under paging conditions. 

### Type of change
Bug fix

### How risky is the change?
Low 

### Backward compatibility
Non breaking change

redknightlois commented on src/Corax/Queries/SortingMatch/SortingMatch.cs:
```diff
@@ -247,9 +281,169 @@ void AddDocumentBoost(ref SortingMatch<TInner, TComparer> match, Span<long> matc
+
+            // We have a take statement so we are only going to care about the highest priority elements. 
+            if (match._take > 0)
+                totalMatches = Math.Min(totalMatches, match._take);
```
@maciejaszyk this will get executed and restrict the amount of actual values to return. The reason is that query builder is assuming that no filtering at Corax level happens but it is setting the take value to perform it. This will cause tests like `CanUseFilterQueryOnMapIndexes` to fail when in reality the query is not setting the `take` parameter.

Query in question:

```
var emp = s.Advanced.RawQuery<Employee>("from Employees order by Name filter Active = false")
    .Statistics(out var stats)
    .SingleOrDefault();
```

Can you take a look why this is happening?

redknightlois committed:
RavenDB-19718: Reimplementation of SortingMatch

redknightlois commented:
@maciejaszyk Check if this doesn't conflict too much what your boosting work.

maciejaszyk commented:
> @maciejaszyk Check if this doesn't conflict too much what your boosting work.

No conflicts for now :)
BTW: a lot of tests failed

redknightlois committed:
RavenDB-19718: Remove the usage of take parameter at the Corax primitives level.

We will allow RavenDB handle that internally for now, there is no need to limit at the Corax level because pagination will break. However, it is certainly a very important optimization that would have to be reintroduced. See https://issues.hibernatingrhinos.com/issue/RavenDB-19738/Reintroduce-Take-operation-at-the-Corax-level.

redknightlois commented:
If all tests passes this is ready to be merged. 

> >@maciejaszyk Check if this doesn't conflict too much what your boosting work.

> No conflicts for now :) BTW: a lot of tests failed

Yes, it was not ready and the modification we did the other day around `Distinct` and other operations trips over assumptions everywhere. Opened [RavenDB-19738](https://issues.hibernatingrhinos.com/issue/RavenDB-19738/Reintroduce-Take-operation-at-the-Corax-level.) to track this particular case. No need to deal with that for now.

redknightlois committed:
RavenDB-19718: Reimplementation of the SortingMultiMatch

redknightlois committed:
RavenDB-19718: Fixed the MultiMatchSorting test.

redknightlois commented:
@arekpalinski How this test worked before I have no idea. The test itself is incorrect, it was ordering **_descending then ascending_** instead of **_ascending then descending_**.

redknightlois commented on src/Corax/Queries/SortingMatch/Comparers/SortingMatch.AscendingMatchComparer.cs:
```diff
@@ -23,6 +23,9 @@ public AscendingMatchComparer(IndexSearcher searcher, OrderMetadata orderMetadat
 
+                if (orderMetadata.Ascending == false)
+                    throw new ArgumentException($"The metadata for field '{orderMetadata.Field.FieldName}' is not marked as 'Ascending == true' ");
+
```
@maciejaszyk We have another API inconsistency here. We can request an `AscendingMatchComparer` with order metadata saying that it is descending. I added a check to fail-fast if that happens for now.

arekpalinski commented:
Please check failing tests

redknightlois commented:
OK. Interesting development on this front. At least one of the failed tests (this is the one I investigated) `SortingTest` has a very interesting behavior. One would suppose that the latest commit would be wrong somehow mainly because if you take it out, the fail is not reproduced (it returns 10000 results). BUT, if you run it under the debugger without the commit, it will fail returning just 1664 results. 

The funny part is that you can actually trigger it straight from a console application. In here I am running both, both in Debug but one without the debugger attached. 
![image](https://user-images.githubusercontent.com/1045411/211684912-3b0852b7-792b-48c6-bedc-ee9d5b4a8c58.png)

If anyone can explain this to me, I would be grateful... I don't know what else to do. 

Could be this related to the bug @maciejaszyk found on the posting list?

ayende commented:
Different code gen if the debugger is attached?

redknightlois commented:
Shouldnt be the case. No. Debuggers use hardware traps afaik, never built
one though.

On Wed, Jan 11, 2023, 3:58 AM Ayende Rahien ***@***.***>
wrote:

> Different code gen if the debugger is attached?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/pull/15533#issuecomment-1378313877>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH7HIYRTY64Q27M5UWNSQ3WRZKXPANCNFSM6AAAAAATPB2MNQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ayende commented:
The *JIT* may decide to generate different code if the debugger is attached

redknightlois committed:
RavenDB-19718: Fixes a race condition in the test.

redknightlois commented:
Race condition on the test. With debugger attached, indexing is slower causing to see consistently a different number.
